### PR TITLE
Keep store in sync with search params and vice versa

### PR DESCRIPTION
**Pull request by Joozty**
@viktor-yakubiv This PRs show proof of concept of keeping URL and store in sync. I wanted to keep only one way either build a state from URL or build URL from the state but that wouldn't suit our use case. Therefore I decided to develop this feature. I aimed to keep everything in one place and that place became store. 

It works like the following:
1. Store is built from params
2. Params are validated and converted if needed. For example, URLSearch object keeps everything as a string and therefore there is a need for conversion
3. If params are updated then URL is updated as well
4. If the URL is changed store is updated too.

I am very excited about this feature and at the same time, I am scared as it may bring bugs hard to debug. However, I think we need it. 

The rationale behind this PR is that I want to use next/link in pagination and therefore I needed to update store from URL.  I wanted to keep the page as parsable as possible and buttons are not a good way to go because [google doesn't click on them](https://support.google.com/webmasters/answer/9112205).

### CODE REVIEW

File: hooks/use-sync-url-params-with-store/index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    const newParams = Array.from(params.entries()).filter(
+      ([, value]) => value != null
+    )
+
+    const query = Object.fromEntries(newParams)
+
+    router.push(
+      { pathname: url, query },
+      { pathname: as, query },
+      { shallow: true }
```
Comment (viktor-yakubiv) [inline hooks/use-sync-url-params-with-store/index.js:16]: If you use 'shallow' you may just get a URL from the router and don't pass it as a parameter.

File: hooks/use-sync-url-params-with-store/index.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    )
+  }
+
+  const handleRouteChange = (nextUrl) => {
```
Comment (viktor-yakubiv) [inline hooks/use-sync-url-params-with-store/index.js:20]: I would add useCallback here to subscribe safely if the page rerenders

**Review by viktor-yakubiv — COMMENTED** [review]
It looks very good! Great job! 👍

I am not sure if you want me to study the code deeper but it's a very good idea and improvement from the first look.

File: hooks/use-sync-url-params-with-store/index.js (hunk 1 above)
Comment (Joozty) [reply hooks/use-sync-url-params-with-store/index.js:16]: Yes, that's something I didn't test yet.  Will probably do it as you are suggesting.

**Comment by viktor-yakubiv** [comment]
Oh, we reached the limit again. Perhaps, many packages to update.

**Comment by Joozty** [comment]
No need to study code deeply. Just keep in mind that if you run in any inconsistency it's most likely this feature. Just notify me but I believe it will work like a charm. :)

**Pull request merged**